File: bff/src/domain/usecases/get-pokemons-by-type.usecase.ts
import { Pokemon } from '../entities/pokemon.entity';
import { PokemonRepository } from '../repositories/pokemon.repository';

export class GetPokemonsByTypeUseCase {
  constructor(private readonly pokemonRepository: PokemonRepository) {}

  async execute(type: string): Promise<Pokemon[]> {
    return this.pokemonRepository.getPokemonsByType(type);
  }
}

File: app/lib/features/pokemon/domain/usecases/get_pokemons_by_type_usecase.dart
import '../../../../core/domain/errors/result.dart';
import '../entities/pokemon_entity.dart';
import '../repositories/i_pokemon_repository.dart';

class GetPokemonsByTypeUseCase {
  final IPokemonRepository _repository;

  GetPokemonsByTypeUseCase(this._repository);

  Future<Result<List<PokemonEntity>>> call(String type) async {
    return await _repository.getPokemonsByType(type);
  }
}
